### ISSUE TITLE
Support for default methods

### ISSUE THREAD

**Issue by ronald-d-rogers** [cbeust/testng#966]
Currently tests defined as default methods aren't run. These are particularly useful for cutting down code duplication for partially horizontal concerns. For example, for testing "widgets" in selenium, which may be part of some of the pages' test classes but not all of them, you can just implement some kind of IHas*WidgetName* interface containing tests for it.

The problem AFAIK is that the default ITestMethodFinder doesn't pick up the methods as it uses Class.getDeclaredMethods() internally, which does not include them.

https://github.com/cbeust/testng/blob/46017205fd97839302fdee7f67265132cdb7ed1b/src/main/java/org/testng/internal/TestNGMethodFinder.java#L45-L49

https://github.com/cbeust/testng/blob/46017205fd97839302fdee7f67265132cdb7ed1b/src/main/java/org/testng/internal/annotations/AnnotationHelper.java#L220

I have some working preliminary code which I would be happy to create a pull request for for more discussion:
https://github.com/ronald-d-rogers/testng/commit/94f4860ec102d15ee98879db48958bbcdd9884e3

**Comment by cbeust** [comment]
I don't want to make Java 8 mandatory for TestNG yet, so your solution would have to be backward compatible. I'm not quite sure how to do that besides using reflection, or do you have a better idea?

**Comment by ronald-d-rogers** [comment]
I tested the solution with Java 7 and it seemed to build and run fine. It is still using only Java 7 reflection. It just checks if the methods on the implementing interfaces are abstract or not. It seems to pass all of the 1k or so tests that ran for me as well. There are some other problems that I saw while cursorily poking through the code, like MethodHelper not being able to find the methods as it also just uses getDeclaredMethods(), but it seems that can pretty easily be changed if it is a problem.
